astropy/io/fits/hdu/image.py

```diff
@@ -135,9 +147,31 @@ def __init__(self, data=None, header=None, do_not_scale_image_data=False,
                 self._data_needs_rescale = True
             return
         else:
+            # Setting data will set set _bitpix, _bzero, and _bscale to the
+            # appropriate BITPIX for the data, and always sets _bzero=0 and
+            # _bscale=1.
             self.data = data
             self.update_header()
 
+            # Check again for BITPIX/BSCALE/BZERO in case they changed when the
+            # data was assigned. This can happen, for example, if the input
+            # data is an unsigned int numpy array.
+            self._bitpix = self._header.get('BITPIX')
+
+            # Do not provide default values for BZERO and BSCALE here because
+            # the keywords will have been deleted in the header if appropriate
+            # after scaling. We do not want to put them back in if they
+            # should not be there.
+            self._bzero = self._header.get('BZERO')
+            self._bscale = self._header.get('BSCALE')
```

Comment:
Ideally, (and I'm not saying it should be done for this PR, but sooner rather than later) this should be inverted.  What's going on is that in some places bzero/bscale are being handled through the header.  For example `update_header()` calls `_update_uint_scale_keywords()` which in turn sets the correct BSCALE/BZERO for unsigned int data *in the header*.  `_update_header_scale_info()` then deletes those keywords again and messes with `self._bscale` and self._bzero`.

Basically what I'm saying is the header should be ignored as early as possible when it comes to managing the state of these attributes.  The bzero and bscale should be managed *only* through the attributes `self._bzero, self._bscale, self._orig_bzero, self._orig_bscale` (and likewise for bitpix).  

self._bzero and self._bzero should at all times contain the values that *should* go in the header if the header were updated to reflect the current data (which it should be when self.data is set).  For unsigned ints this means `self._bscale = 1` and `self._bzero = <offset>`.  For floats it typically means `bzero=0, bscale=1` unless the data has been explicitly rescaled with `scale()`.  Finally `update_header()` should only have to look at the current values of `self._bscale` and `self._bzero` to update those keywords (as it does `self._bitpix`).

`_orig_bscale` and `_orig_bzero` should only be used for keeping track of the original values they had when the HDU is first initialized with a header (or set to sensible defaults when they were not in the original header).  I think this is already fairly consistently the case.  

Forgive me if all of the above is obvious to you already. I'm just trying to sort out my own thoughts.  This PR looks like a step in the right direction, but there's still a bit of a mess involving management of these attributes that makes the code more complicated and confusing than it needs to be.  Not saying that needs to be addressed right this second.